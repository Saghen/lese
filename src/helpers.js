import styled, { css } from "styled-components";

export function arrayToCSS(array) {
  return `${array.join(";")};`;
}

export function getFlexProperties({
  column,
  xAlign,
  yAlign,
  wrap,
  separation
}) {
  const properties = [];
  properties.push("display: flex");

  // Set defaults
  xAlign = xAlign && returnDefault(xAlign, "string", "center");
  yAlign = yAlign && returnDefault(yAlign, "string", "center");
  wrap = wrap && returnDefault(wrap, "string", "wrap");

  if (wrap) properties.push(`flex-wrap: ${wrap}`);

  if (column) properties.push("flex-direction: column");

  if (xAlign)
    properties.push(
      column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    );
  if (yAlign)
    properties.push(
      !column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    );

  if (separation)
    properties.push(
      `
        > * {
          margin: 0;
        }
        > * + * {
          ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
        }
      `
    );

  arrayToCSS(properties);
}

export function getTextProperties({
  color,
  fontSize,
  fontDecoration,
  fontWeight
}) {
  const properties = [];
  if (fontSize) properties.push(`font-size: ${fontSize}`);
  if (fontDecoration) properties.push(`font-decoration: ${fontDecoration}`);
  if (fontWeight) properties.push(`font-weight: ${fontWeight}`);

  return arrayToCSS(properties);
}

export function getSizeProperties({
  height,
  width,
  responsive,
  padding,
  margin
}) {
  const properties = [];
  if (width) {
    if (responsive) properties.push(`max-width: ${width}`, `width: 100%`);
    else properties.push(`width: ${width}`);
  }
  if (height) {
    if (responsive) properties.push(`max-height: ${height}`, `height: 100%`);
    else properties.push(`height: ${height}`);
  }
  return arrayToCSS(properties);
}

function returnDefault(property, type, defaultValue) {
  return typeof property === type ? property : defaultValue;
}

export function getLayoutProperties({ padding, margin, position }) {
  const properties = [];
  if (padding) properties.push(`padding: ${padding}`);
  if (margin) properties.push(`margin: ${margin}`);

  position = returnDefault(position, "string", "relative");
  if (position) properties.push(`position: ${position}`);
  return arrayToCSS(properties);
}

// TODO: Rename this function to something other than "Colouring"
export function getColouringProperties({
  background,
  backgroundColor,
  backgroundImage,
  color
}) {
  const properties = [];
  if (color) properties.push(`color: ${color}`);
  if (background) properties.push(`background: ${background}`);
  if (backgroundColor) properties.push(`background-color: ${backgroundColor}`);
  if (backgroundImage) properties.push(`background-image: ${backgroundImage}`);

  return arrayToCSS(properties);
}
