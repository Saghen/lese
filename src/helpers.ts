function arrayToCSS(array: string[]) {
  return `${array.join(";")};`;
}

function returnDefault(property: any, defaultValue: any) {
  return typeof property === "string" ? property : defaultValue;
}

function camelToKebab(string: string) {
  return string.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
}

export type PropertyGeneratorHandler<T> = (props: T) => string;

export interface PropertyGeneratorOptions<T> {
  default?: string;
  handler?: PropertyGeneratorHandler<T>;
  property?: string;
}

export type PropertyGeneratorKey<T> = Array<[Extract<keyof T, string>, PropertyGeneratorOptions<T> | PropertyGeneratorHandler<T>] | Extract<keyof T, string>>;

export function propertyGenerator<T>(keys: PropertyGeneratorKey<T>): (props: T) => string {
  return (props: T) => {
    const properties = [];
    for (const key of keys) {
      // Exits
      if (Array.isArray(key) && !props[key[0]]) continue;
      if (typeof key === "string" && !props[key]) continue;

      if (!Array.isArray(key)) {
        props[key] && properties.push(`${camelToKebab(key)}: ${props[key]}`);
        continue;
      }

      const isFunction = typeof key[1] === "function";
      const isObject = typeof key[1] === "object";

      if (!(isFunction || isObject))
        throw new TypeError(`Invalid options provided at key: ${key[0]}`);

      // All validated
      if (isFunction) properties.push((key[1] as PropertyGeneratorHandler<T>)(props));

      if (isObject) {
        const options = key[1] as PropertyGeneratorOptions<T>;
        const property = options.property || camelToKebab(key[0]);
        if (options.default) props[key[0]] = returnDefault(props[key[0]], options.default);

        if (typeof options.handler === "function") properties.push(options.handler(props));
        else properties.push(`${property}: ${props[key[0]]}`);
      }
    }
    return arrayToCSS(properties);
  };
}
