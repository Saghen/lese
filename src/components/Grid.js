import styled, { css } from "styled-components";
import Base from "./Base";

import { getGridProperties } from "../helpers";

export default styled(Base)`
  ${getGridProperties}
`;
