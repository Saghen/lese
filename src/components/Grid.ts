import styled from "@emotion/styled";
import Base, { BaseProps } from "./Base";
import { propertyGenerator } from "../helpers";

export interface GridProps extends BaseProps {
  columns?: string;
  rows?: string;
  autoColumns?: string;
  autoRows?: string;
  columnGap?: string;
  rowGap?: string;
  gap?: string;
  xAlign?: boolean | string;
  yAlign?: boolean | string;
  align?: boolean | string;
}

export interface GridChildProps {
  columnSelf?: string;
  rowSelf?: string;
}

const getGridProperties = propertyGenerator<GridProps>([
  ["columns", { property: "grid-template-columns" }],
  ["rows", { property: "grid-template-rows" }],
  ["autoColumns", { property: "grid-auto-columns" }],
  ["autoRows", { property: "grid-auto-rows" }],
  ["columnGap", { property: "grid-column-gap" }],
  ["rowGap", { property: "grid-row-gap" }],
  ["gap", { property: "grid-gap" }],
  ["xAlign", { property: "justify-items", default: "center" }],
  ["yAlign", { property: "align-items", default: "center" }],
  ["align", { property: "place-items", default: "center center" }],
]);

export default styled(Base)<React.PropsWithChildren<GridProps>>`
  display: grid;
  ${getGridProperties}
`;
