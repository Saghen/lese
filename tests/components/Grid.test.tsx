import React from "react";
import Grid from "../../src/components/Grid";
import { matchers } from "@emotion/jest";
import { renderComponent } from "./helpers";

expect.extend(matchers);

test("grid layout props", () => {
  expect(
    renderComponent(<Grid columns="auto auto" rows="auto auto" columnGap="8px" rowGap="8px" gap="8px" />)
  ).toMatchSnapshot();
  expect(renderComponent(<Grid autoColumns="auto" autoRows="auto" />)).toMatchSnapshot();
  expect(
    renderComponent(
      <Grid xAlign yAlign align>
        <div>Test</div>
      </Grid>
    )
  ).toMatchSnapshot();
  expect(
    renderComponent(
      <Grid xAlign="stretch" yAlign="end" align="normal start">
        <div>Test</div>
      </Grid>
    )
  ).toMatchSnapshot();
});
