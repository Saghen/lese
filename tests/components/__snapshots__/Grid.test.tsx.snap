// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid layout props 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto auto;
  grid-template-rows: auto auto;
  grid-column-gap: 8px;
  grid-row-gap: 8px;
  grid-gap: 8px;
}

<div
  className="emotion-0"
  rows="auto auto"
/>
`;

exports[`grid layout props 2`] = `
.emotion-0 {
  display: grid;
  grid-auto-columns: auto;
  grid-auto-rows: auto;
}

<div
  className="emotion-0"
/>
`;

exports[`grid layout props 3`] = `
.emotion-0 {
  display: grid;
  justify-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  place-items: center center;
}

<div
  className="emotion-0"
>
  <div>
    Test
  </div>
</div>
`;

exports[`grid layout props 4`] = `
.emotion-0 {
  display: grid;
  justify-items: stretch;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  place-items: normal start;
}

<div
  className="emotion-0"
>
  <div>
    Test
  </div>
</div>
`;
